test/sh_do_test.sh

```diff
@@ -30,14 +30,36 @@ output_file=${RESULTS_DIR}/${script_name}.out
 echo " ... ${script_name}"
 rm -f ${output_file}
 
-./${script_name} > ${output_file}  2>&1
-if [ $? -ne 0 ]; then
-    # duplicate d_do_test output
-    echo >> ${output_file}
-    echo ============================== >> ${output_file}
-    echo Test ${script_name} failed >> ${output_file}
-
-    echo Test ${script_name} failed. The logged output:
-    cat ${output_file}
-    exit 1
-fi
+function finish {
+    # reset output stream
+    set +x
+    exec 1>&40
+    exec 2>&41
+
+    if [ $1 -ne 0 ]; then
+        echo ==============================
+        echo Test ${script_name} failed. The logged output:
+        cat ${output_file}
+
+        echo ==============================
+        echo Test ${script_name} failed. The xtrace output:
+        cat ${output_file}.log
+
+        rm -rf ${output_file}
+        exit $?
+    fi
+}
+trap 'finish $?' INT TERM EXIT
+
+# redirect stdout + stderr to the output file + keep a reference to the std{out,err} streams for later
+exec 40>&1
+exec 41>&2
```

Comment:
bash has a syntax to assign a free file descriptor and store it in a variable (`exec {name_of_fd_var}>&1`), which should guarantee that there will not be collisions, but the feature is "relatively" new and not sure if it's available on all test systems.

Reply:
Nice! According to [this SO answer](https://stackoverflow.com/a/41620630/3026245) it's available since available since bash 4.1+ (2009-12-31), so I just gave it a try.
Who knows, maybe it works ...